eosdis-nasa/earthdata-pub-api
# WIP: Edpub 1256 note visability backend

# Description

Set up backend code for handling note scope visibility

## Spec

See Ticket: https://bugs.earthdata.nasa.gov/browse/EDPUB-1256

---

## Validation

1. Make sure all merge request checks have passed (CI/CD).
2. Pull related branches locally.
3. In an Incognito browser Create an account with DAAC Data Manager Permissions
4. Log out and Create another account with DAAC Staff permissions
5. Log out and Create another account with Data Producer permissions 
6. Create a submission & Submit
7. Log out and log in with the  DAAC Staff role
8. On the DAR Approval page add comment "DS & DM Only!"
9. Navigate to the conversations tab, click on the conversation for your request and add a note "DM ONLY"
10. Add an additional note "E for everyone"
12. Using inspect element check the bearer token of any api call and take a note of the authentication
13. In your browser navigate to http://localhost:8080/docs/index.html
14. Click the Authorize button in the upper right side of the swagger page and add your bearer token
15. In the Swagger navigate down to the Notification section
16. In your db visualizer of choice get the user id for the dp role and the note ID for the "DM ONLY" note
17. Under /api/notification/add_viewers run the command with the following
```
# {"viewer_ids":["<dp_user_id>"], "note_id": "<note_id>"}
# I.E
{"viewer_ids":["377d255d-366a-4ef5-bc34-18f7514ee107"], "note_id": "f434e423-dc40-41ce-a991-a693840f63af"}
```
18. Verify that the user Id was added to the note_scope table for that note_id
19. Under /api/notification/remove_viewer run the command with the following
```
# {"viewer_id":"<dp_user_id>", "note_id": "<note_id>"}
# I.E
{"viewer_id":"377d255d-366a-4ef5-bc34-18f7514ee107", "note_id": "f434e423-dc40-41ce-a991-a693840f63af"}
```
20. Verify that the user Id was removed from the note_scope table for that note_id
21. In your db visualizer of choice get the role ID for the DAAC Data Manager role and DAAC Staff role and the note ID for the "DM ONLY" note
22. Under /api/notification/add_viewer_roles run the command with the following
```
# {"viewer_roles":["<role_id>"], "note_id": "<note_id>"}
# I.E
{"viewer_roles": ["fb348f27-6604-441b-b86c-8a39876b6bd0", "ca1d169f-db41-462b-9204-996447f7fa39" ], "note_id": "f434e423-dc40-41ce-a991-a693840f63af"}
```
23. Verify that the roles were added to the role list in the note_scope table for that note_id
24.  In your db visualizer of choice get the role ID for the DAAC Data Manager role and DAAC Staff role and the note ID for the "DS & DM Only!" note
25. Under /api/notification/add_viewer_roles run the command with the following
```
# {"viewer_roles":["<role_id>"], "note_id": "<note_id>"}
# I.E
{"viewer_roles": ["fb348f27-6604-441b-b86c-8a39876b6bd0", "ca1d169f-db41-462b-9204-996447f7fa39" ], "note_id": "b1a6d097-d6a9-4f13-87e2-1bffb3cee7d0"}
```
26. Verify that the roles were added to the role list in the note_scope table for that note_id
27. Under /api/notification/remove_viewer_roles run the command to remove the DAAC Staff role
```
# {"viewer_roles":["<role_id>"], "note_id": "<note_id>"}
# I.E
{"viewer_role": "ca1d169f-db41-462b-9204-996447f7fa39" , "note_id": "f434e423-dc40-41ce-a991-a693840f63af"}
```
28. Verify that the role was removed from the role list in the note_scope table for that note_id
29 Log as the DAAC Manager role and navigate to the conversations tab. Ensure that all conversations expected to appear do
30. Log in as the DAAC Staff role and navigate to the conversations tab. Ensure that all conversations expected to appear do
31. Look at the DAR Approval page and ensure the appropriate notes appear there
32. Log in as the Data Producer role and navigate to the conversations tab. Ensure that all conversations expected to appear do
